File: src/main/java/com/gammon/pcms/service/VariationKpiService.java
package com.gammon.pcms.service;

import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.gammon.pcms.dao.VariationKpiRepository;
import com.gammon.pcms.model.VariationKpi;

import edu.emory.mathcs.backport.java.util.Arrays;

@Transactional
@Service
public class VariationKpiService {

	private Logger logger = LoggerFactory.getLogger(getClass());
	
	@Autowired
	private VariationKpiRepository repository;
	
	public List<VariationKpi> findByJobNo(String jobNo) {
		return repository.findByNoJob(jobNo);
	}
	
	public List<VariationKpi> getByYear(String jobNo, int year) {
		return repository.getByYear(jobNo, year);
	}
	
	public VariationKpi save(String jobNo, VariationKpi kpi) {
		kpi.setNoJob(jobNo);
		return repository.save(kpi);
	}
	
	public List<VariationKpi> save(String jobNo, List<VariationKpi> kpiList) {
		kpiList.forEach(kpi -> {
			kpi.setNoJob(jobNo);
		});
		return repository.save(kpiList);
	}
	
	public void delete(String jobNo, Long[] ids) {
		List<Long> idList = Arrays.asList(ids);
		idList.forEach(id -> {
			VariationKpi kpiDb = repository.getOne((Long)id);
			if(kpiDb.getNoJob().equals(jobNo)) {
				repository.delete(kpiDb);
			} else {
				throw new IllegalAccessError("cannot delete variation kpi");
			}
		});
	}
	
	public Page<VariationKpi> filterPagination(
			Pageable pageable,
			String noJob,
			String year, 
			String month, 
			String numberIssued, 
			String amountIssued, 
			String numberSubmitted,
			String amountSubmitted,
			String numberAssessed,
			String amountAssessed,
			String numberApplied,
			String amountApplied,
			String numberCertified,
			String amountCertified,
			String remarks) {
		return repository.filterPagination(
				pageable, 
				noJob, 
				year  + "%", 
				month + "%", 
				numberIssued  + "%", 
				amountIssued  + "%", 
				numberSubmitted  + "%", 
				amountSubmitted  + "%", 
				numberAssessed  + "%", 
				amountAssessed  + "%", 
				numberApplied  + "%", 
				amountApplied  + "%", 
				numberCertified  + "%", 
				amountCertified  + "%", 
				"%" + remarks + "%"
				);
	}
}

File: src/main/webapp/js/service/variation-kpi-service.js
mainApp.service('variationKpiService', ['$http', '$q', '$log', 'GlobalHelper',  function($http, $q, $log, GlobalHelper){
	// Return public API.
    return({
    	findByJobNo:	findByJobNo,
    	saveByJobNo: 	saveByJobNo,
    	saveListByJobNo: saveListByJobNo,
    	deleteByJobNo:	deleteByJobNo,
    	getByPage:		getByPage
    });
	
    function findByJobNo(jobNo) {
		var request = $http({
			method: "get",
			url: "service/variationKpi/" + jobNo
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }

    function getByPage(page, size, direction, property,
    		noJob, 
			year, 
			month, 
			numberIssued, 
			amountIssued, 
			numberSubmitted, 
			amountSubmitted, 
			numberAssessed, 
			amountAssessed, 
			numberApplied, 
			amountApplied, 
			numberCertified, 
			amountCertified, 
			remarks
		) {
		var request = $http({
			method: "get",
			url: "service/variationKpi/" + page + "/" + size,
			params: {
				direction: direction,
				property: property,
				noJob: noJob, 
				year: year, 
				month: month, 
				numberIssued: numberIssued, 
				amountIssued: amountIssued, 
				numberSubmitted: numberSubmitted, 
				amountSubmitted: amountSubmitted, 
				numberAssessed: numberAssessed, 
				amountAssessed: amountAssessed, 
				numberApplied: numberApplied, 
				amountApplied: amountApplied, 
				numberCertified: numberCertified, 
				amountCertified: amountCertified, 
				remarks: remarks
			}
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }
    
    function saveByJobNo(jobNo, kpi) {
		var request = $http({
			method: "post",
			url: "service/variationKpi/" + jobNo,
			data: kpi
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }
    
    function saveListByJobNo(jobNo, kpiList) {
		var request = $http({
			method: "post",
			url: "service/variationKpi/saveList/" + jobNo,
			data: kpiList
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }
    
    function deleteByJobNo(jobNo, id) {
		var request = $http({
			method: "delete",
			url: "service/variationKpi/" + jobNo + "/" + id
		});
		return( request.then( GlobalHelper.handleSuccess, GlobalHelper.handleError ) );
    }
}]);